File: frontend/src/app/core/auth.service.ts
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { ApiService } from './api.service';
import { AuthResponse, User, UserRole } from './models';

@Injectable({ providedIn: 'root' })
export class AuthService {
  private readonly userSubject = new BehaviorSubject<User | null>(this.readUser());
  readonly user$ = this.userSubject.asObservable();

  constructor(private api: ApiService) {}

  get user(): User | null {
    return this.userSubject.value;
  }

  get token(): string | null {
    return localStorage.getItem('kk_token');
  }

  get isLoggedIn(): boolean {
    return !!this.token && !!this.user;
  }

  get isProducer(): boolean {
    const role = this.user?.role;
    return role === 'PRODUCER' || role === 'BOTH';
  }

  async register(payload: {
    name: string;
    email: string;
    password: string;
    role: UserRole;
    city?: string;
    latitude?: number;
    longitude?: number;
  }): Promise<User> {
    const res = await this.api.request<AuthResponse>('/api/auth/register', {
      method: 'POST',
      body: JSON.stringify(payload),
    }, false);
    this.persist(res);
    return res.user;
  }

  async login(email: string, password: string): Promise<User> {
    const res = await this.api.request<AuthResponse>('/api/auth/login', {
      method: 'POST',
      body: JSON.stringify({ email, password }),
    }, false);
    this.persist(res);
    return res.user;
  }

  /** MVP Google login stub — wire Firebase/Google Identity later */
  async googleLogin(payload: { idToken: string; email: string; name: string; role?: UserRole }): Promise<User> {
    const res = await this.api.request<AuthResponse>('/api/auth/google', {
      method: 'POST',
      body: JSON.stringify(payload),
    }, false);
    this.persist(res);
    return res.user;
  }

  async refreshMe(): Promise<User> {
    const user = await this.api.request<User>('/api/auth/me');
    localStorage.setItem('kk_user', JSON.stringify(user));
    this.userSubject.next(user);
    return user;
  }

  async updateProfile(payload: Record<string, unknown>): Promise<User> {
    const user = await this.api.request<User>('/api/auth/me', {
      method: 'PUT',
      body: JSON.stringify(payload),
    });
    localStorage.setItem('kk_user', JSON.stringify(user));
    this.userSubject.next(user);
    return user;
  }

  logout(): void {
    localStorage.removeItem('kk_token');
    localStorage.removeItem('kk_user');
    this.userSubject.next(null);
  }

  private persist(res: AuthResponse): void {
    localStorage.setItem('kk_token', res.token);
    localStorage.setItem('kk_user', JSON.stringify(res.user));
    this.userSubject.next(res.user);
  }

  private readUser(): User | null {
    const raw = localStorage.getItem('kk_user');
    if (!raw) {
      return null;
    }
    try {
      return JSON.parse(raw) as User;
    } catch {
      return null;
    }
  }
}

File: mobile/lib/services/auth_service.dart
import 'dart:convert';

import 'package:flutter/foundation.dart';
import 'package:lokal/models/models.dart';
import 'package:lokal/services/api_client.dart';
import 'package:shared_preferences/shared_preferences.dart';

class AuthService extends ChangeNotifier {
  AuthService(this._api);

  final ApiClient _api;
  User? user;
  bool ready = false;

  bool get isLoggedIn => user != null && _api.token != null;
  bool get isProducer => user?.isProducer ?? false;

  Future<void> bootstrap() async {
    final prefs = await SharedPreferences.getInstance();
    final token = prefs.getString('lokal_token');
    final rawUser = prefs.getString('lokal_user');
    if (token != null && rawUser != null) {
      _api.token = token;
      user = User.fromJson(jsonDecode(rawUser) as Map<String, dynamic>);
      try {
        await refreshMe();
      } catch (_) {
        await logout();
      }
    }
    ready = true;
    notifyListeners();
  }

  Future<void> register({
    required String name,
    required String email,
    required String password,
    required UserRole role,
    double? latitude,
    double? longitude,
    String? city,
  }) async {
    final res = await _api.request<Map<String, dynamic>>(
      'POST',
      '/api/auth/register',
      auth: false,
      body: {
        'name': name,
        'email': email,
        'password': password,
        'role': roleApi(role),
        'latitude': latitude,
        'longitude': longitude,
        'city': city,
      },
    );
    await _persist(res);
  }

  Future<void> login(String email, String password) async {
    final res = await _api.request<Map<String, dynamic>>(
      'POST',
      '/api/auth/login',
      auth: false,
      body: {'email': email, 'password': password},
    );
    await _persist(res);
  }

  Future<void> googleDemo() async {
    final res = await _api.request<Map<String, dynamic>>(
      'POST',
      '/api/auth/google',
      auth: false,
      body: {
        'idToken': 'demo-google-token',
        'email': 'google.user.${DateTime.now().millisecondsSinceEpoch}@gmail.com',
        'name': 'Google Kasutaja',
        'role': 'BUYER',
      },
    );
    await _persist(res);
  }

  Future<void> refreshMe() async {
    final json = await _api.request<Map<String, dynamic>>('GET', '/api/auth/me');
    user = User.fromJson(json);
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString('lokal_user', jsonEncode(json));
    notifyListeners();
  }

  Future<void> updateProfile(Map<String, dynamic> body) async {
    final json = await _api.request<Map<String, dynamic>>('PUT', '/api/auth/me', body: body);
    user = User.fromJson(json);
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString('lokal_user', jsonEncode(json));
    notifyListeners();
  }

  Future<void> logout() async {
    _api.token = null;
    user = null;
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove('lokal_token');
    await prefs.remove('lokal_user');
    notifyListeners();
  }

  Future<void> _persist(Map<String, dynamic> res) async {
    final token = res['token'] as String;
    final userJson = res['user'] as Map<String, dynamic>;
    _api.token = token;
    user = User.fromJson(userJson);
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString('lokal_token', token);
    await prefs.setString('lokal_user', jsonEncode(userJson));
    notifyListeners();
  }
}
